scripts/ai/jobs/CpAIJobFieldWork.lua

```diff
@@ -85,46 +90,35 @@ function CpAIJobFieldWork:applyCurrentState(vehicle, mission, farmId, isDirectSt
     else
         local x, z = self.cpJobParameters.fieldPosition:getPosition()
         if x == nil or z == nil then
+            -- if there is no field position set, set it to the vehicle's position
             x, _, z = getWorldTranslation(vehicle.rootNode)
             self.cpJobParameters.fieldPosition:setPosition(x, z)
         end
     end
-end
-
---- Checks the field position setting.
-function CpAIJobFieldWork:validateFieldSetup(isValid, errorMessage)
-    if not isValid then
-        return isValid, errorMessage
+    local fieldPolygon = vehicle:cpGetFieldPolygon()
+    if fieldPolygon then
+        -- if we already have a field polygon, show it
+        self.selectedFieldPlot:setWaypoints(fieldPolygon)
+        self.selectedFieldPlot:setVisible(true)
     end
+end
 
-    local vehicle = self.vehicleParameter:getVehicle()
-
-    -- everything else is valid, now find the field
-    local tx, tz = self.cpJobParameters.fieldPosition:getPosition()
-    if tx == nil or tz == nil then
-        return false, g_i18n:getText("CP_error_not_on_field")
-    end
-    self.hasValidPosition = false
-    self.foundVines = nil
-    local fieldPolygon
-    fieldPolygon, self.isCustomField = CpFieldUtil.getFieldPolygonAtWorldPosition(tx, tz)
-    self:setFieldPolygon(fieldPolygon)
+function CpAIJobFieldWork:onFieldBoundaryDetectionFinished(vehicle, fieldPolygon, islandPolygons)
     if fieldPolygon then
-        self.hasValidPosition = true
-        self.foundVines = g_vineScanner:findVineNodesInField(fieldPolygon, tx, tz, self.customField ~= nil)
+        local x, z = vehicle:cpGetFieldPosition()
+        self.foundVines = g_vineScanner:findVineNodesInField(fieldPolygon, x, z, self.customField ~= nil)
         if self.foundVines then
             CpUtil.debugVehicle(CpDebug.DBG_FIELDWORK, vehicle, "Found vine nodes, generating a vine field border.")
-            fieldPolygon = g_vineScanner:getCourseGeneratorVertices(0, tx, tz)
+            fieldPolygon = g_vineScanner:getCourseGeneratorVertices(0, x, z)
         end
         self.selectedFieldPlot:setWaypoints(fieldPolygon)
         self.selectedFieldPlot:setVisible(true)
-        self.selectedFieldPlot:setBrightColor(true)
     else
         self.selectedFieldPlot:setVisible(false)
+        -- TODO: here we need to tell somehow the frame about the detection success/failure
```

Comment:
We could register a callback in the frame or pass a finishing closure, if we want to do it completely asynchron.